README.md
**trezor-wallet /**
[Readme](https://cosmic.plus/#view:js-trezor-wallet)
• [Contributing](https://cosmic.plus/#view:js-trezor-wallet/CONTRIBUTING)
• [Changelog](https://cosmic.plus/#view:js-trezor-wallet/CHANGELOG)

# Readme

![Licence](https://img.shields.io/github/license/cosmic-plus/js-trezor-wallet.svg)
[![Dependencies](https://badgen.net/david/dep/cosmic-plus/js-trezor-wallet)](https://david-dm.org/cosmic-plus/js-trezor-wallet)
![Vulnerabilities](https://snyk.io/test/npm/@cosmic-plus/trezor-wallet/badge.svg)
![Bundle](https://badgen.net/badgesize/gzip/cosmic-plus/js-trezor-wallet-web/master/trezor-wallet.js?label=bundle)
![Downloads](https://badgen.net/npm/dt/@cosmic-plus/trezor-wallet)

Easy Trezor wallet support for Stellar applications.

(Weekly updates: [Reddit](https://reddit.com/r/cosmic_plus),
[Twitter](https://twitter.com/cosmic_plus),
[Keybase](https://keybase.io/team/cosmic_plus),
[Telegram](https://t.me/cosmic_plus))

## Introduction

This library is a convenient wrapper around the official
[TrezorConnect](https://github.com/trezor/connect) library.

It provides a way to support Trezor devices with a few one-liners:

```js
// Step 1: Connect
await trezorWallet.connect()

// Step 2: Get public key
const pubkey = trezorWallet.publicKey

// Step 3: Sign
await trezorWallet.sign(transaction)

// Extra: Event handlers
trezorWallet.onConnect = connectionHandler
trezorWallet.onDisconnect = disconnectionHandler
```

This library is browser-only.

**Known limitations**

A few operations & parameters are not supported yet. An explicit error
message will get thrown if you try to sign a transaction including them:

- Binary text memo containing NULL characters,
- The `manageBuyOffer` operation,
- The `pathPaymentStrictSend` operation.

If you encounter an unexpected error, please [fill an
issue](https://github.com/cosmic-plus/js-trezor-wallet/issues/new/choose).

## Installation

### NPM/Yarn

- NPM: `npm install @cosmic-plus/trezor-wallet`
- Yarn: `yarn add @cosmic-plus/trezor-wallet`

In your script: `const trezorWallet = require("@cosmic-plus/trezor-wallet")`

### Bower

`bower install cosmic-plus-trezor-wallet`

In your HTML page:

```HTML
<script src="./bower_components/cosmic-plus-trezor-wallet/trezor-wallet.js"></script>
```

### CDN

In your HTML page:

```HTML
<script src="https://cdn.cosmic.plus/trezor-wallet@1.x"></script>
```

_Note:_ For production release it is advised to serve your copy of the library.

## Usage

### Functions

#### trezorWallet.register(appUrl, email)

Registers yourself in the Trezor Connect Manifest. This provides them the
ability to reach you in case of any required maintenance.

This subscription is mandatory and the library register the Cosmic.plus
contact by default.

**See**: [Trezor Connect Manifest](https://github.com/trezor/connect/blob/develop/docs/index.md#trezor-connect-manifest)  
**Parameters:**

- `appUrl`: _String_ - Application URL.
- `email`: _String_ - Developer email.

#### await trezorWallet.connect([account])

Waits for a connection with a Trezor wallet. If **account** is not provided,
account 1 is used. The library will stop listening for a connection if
`trezorWallet.disconnect()` is called.

Once the connection is established, you can use `await trezorWallet.connect(account)` again at any time to ensure the device is
still connected.

When switching to another **account**, you can `await trezorWallet.connect(new_account)` without prior disconnection.

_Note:_ To stay consistent with the way Trezor numbers accounts, **account**
starts at 1 (derivation path: `m/44'/148'/0'`).

**Parameters:**

- `account`: _Number | String (default: 1)_ - Either an account number (starts at 1)
  or a derivation path (e.g: `m/44'/148'/0'`).

#### await trezorWallet.sign(transaction)

Prompts the user to accept **transaction** using the connected account of
their Trezor device.

If the user accepts, adds the signature to **transaction**. Else, throws an
error.

**Parameters:**

- `transaction`: _Transaction_ - A StellarSdk Transaction

#### trezorWallet.disconnect()

Close the connection with the Trezor device, or stop waiting for one in case
a connection has not been established yet.

#### trezorWallet.newAccount([horizon])

Connects the first unused account.

_Note:_ merged accounts are considered as used.

**Parameters:**

- `horizon`: _String | Server (default: https://horizon.stellar.org)_ - The
  Horizon server where to check for account existence. It can be either an URL
  or a _StellarSdk.Server_ object.

#### trezorWallet.scan([params]) ⇒ _Array_

Scans the Trezor device for accounts that exist on **params.horizon**. The
scanning stops after encountering **params.attempts** unused accounts.

Merged accounts are considered as existing accounts and will reset the
**params.attempts** counter when encountered.

Returns an _Array_ of _Objects_ containing `account` number, `publicKey`,
`path`, and `state` (either `"open"` or `"merged"`).

**Parameters:**

- `params`: _Object_ - Optional parameters.
  - `horizon`: _String | Server (default: https://horizon.stellar.org)_ - The
    Horizon server where to check for account existence. It can be either an URL
    or a _StellarSdk.Server_ object.
  - `attempts`: _Number (default: 3)_ - The number of empty accounts before
    scanning stops.
  - `includeMerged`: _Boolean (default: false)_ - List merged accounts as well.

#### trezorWallet.getPublicKeys([start], [length]) ⇒ _Array_

Request multiple public keys from the Trezor device. The request will return
**length** accounts, starting by **start** (minimum 1).

Returns an _Array_ of _Objects_ with properties `account`, `publicKey`, and
`path`.

**Parameters:**

- `start`: _Number (default: 1)_ - Starting account number
- `length`: _Number (default: 1)_ - Number of account to be listed.

### Events

#### trezorWallet.onConnect : _function_

_Function_ to execute on each connection.

#### trezorWallet.onDisconnect : _function_

_Function_ to execute on each disconnection.

### Members

#### trezorWallet.publicKey : _String_

PublicKey of the connected account.

#### trezorWallet.path : _String_

Derivation path of the connected account. (default: `m/44'/148'/0'`)

## Links

**Organization:** [Cosmic.plus](https://cosmic.plus/) | [@GitHub](https://git.cosmic.plus) | [@NPM](https://www.npmjs.com/search?q=cosmic-plus)

**Follow:** [Reddit](https://reddit.com/r/cosmic_plus) | [Twitter](https://twitter.com/cosmic_plus) | [Medium](https://medium.com/cosmic-plus) | [Codepen](https://codepen.io/cosmic-plus)

**Talk:** [Telegram](https://t.me/cosmic_plus) | [Keybase](https://keybase.io/team/cosmic_plus)


CHANGELOG.md
**trezor-wallet /**
[Readme](https://cosmic.plus/#view:js-trezor-wallet)
• [Contributing](https://cosmic.plus/#view:js-trezor-wallet/CONTRIBUTING)
• [Changelog](https://cosmic.plus/#view:js-trezor-wallet/CHANGELOG)

# Changelog

All notable changes to this project will be documented in this file.

This project adheres to **[Semantic
Versioning](https://semver.org/spec/v2.0.0.html)**. Version syntax is
`{major}.{minor}.{patch}`, where a field bump means:

- **Patch**: The release contains bug fixes.
- **Minor**: The release contains backward-compatible changes.
- **Major**: The release contains compatibility-breaking changes.

**Remember:** Both micro and minor releases are guaranteed to respect
backward-compatibility and can be updated to without risk of breakage. For major
releases, please check this changelog before upgrading.

## 1.7.0 - 2021-09-17

### Changed

- Meta: Update dependencies.
- Meta: Drop support for unsafe browsers.

## 1.6.0 - 2021-01-11

### Changed

- Logic: Upgrade [stellar-sdk] to 7.x.
- Logic: Update [trezor-connect] to 8.1.19.

## 1.5.0 - 2020-11-08

### Changed

- Logic: Upgrade [trezor-connect] to 8.1.17.

## 1.4.0 - 2020-09-12

### Changed

- Logic: Update [trezor-connect] to 8.1.14.

## 1.3.1 - 2020-05-31

### Fixed

- Logic: Fix handling of null/zero timebounds.

## 1.3.0 - 2020-05-24

### Changed

- Documentation: `manageData` operation is now fully supported.

### Fixed

- Logic: Fix handling of `pathPayment` operations.

## 1.2.0 - 2020-04-18

### Changed

- Logic: Update [trezor-connect]. (firmware compatibility update + bugfixes)

## 1.1.0 - 2020-03-07

### Changed

- Meta: Upgrade to [stellar-sdk] 4.x.

## 1.0.3 - 2020-02-24

### Fixed

- Logic: Fix a wrong sequence issue. (Thanks [@Tyvdh](https://github.com/tyvdh))

## 1.0.2 - 2020-01-18

### Fixed

- Documentation: Update according to 1.x release. (sorry about that!)
- Logic: Update trezor-connect to 8.0.13.
  ([bugfixes](https://github.com/trezor/connect/blob/develop/CHANGELOG.md#8013))

## 1.0.1 - 2019-12-28

### Fixed

- Logic: Fix `.scan()` handling of merged account. (They were stated as opened)

## 1.0.0 - 2019-12-07

### Breaking

- API: Breaking change: account=1 is always the default for `.connect()`.
  Previously, the default was the latest used account number, or 1. This was
  leaving room for errors in account handling and has been changed to a clearer
  behavior. As of now, the right way to ensure the connection is still alive is
  to use `trezorWallet.connect(account)`, except when if _account_ is 1.
- Logic: Breaking change: set `trezor.path` once connected. Before,
  `trezor.path` was set before connection actually happened, which was an
  inconsistent behavior.

## 0.5.0 - 2019-11-23

### Changed

- Logic: Update [trezor-connect] to 8.0.10. (bugfixes)

## 0.4.0 - 2019-11-09

### Added

- API: Add `.getPublicKeys()`. This method retrieves multiple public keys at
  once. ([See
  documentation](https://cosmic.plus/#view:js-trezor-wallet/%23trezorwalletgetpublickeys))

- API: Add `.scan()`. This method scans for existing accounts on the Trezor
  Device. ([See
  documentation](https://cosmic.plus/#view:js-trezor-wallet/%23trezorwalletscan))

- API: Add `.newAccount()`. This method connects the first unused account of a
  Trezor device. ([See
  documentation](https://cosmic.plus/#view:js-trezor-wallet/%23trezorwalletnewaccount))

### Changed

- API: Make `.register()` optional. The library now register using <Cosmic.plus>
  credentials by default.

- Documentation: Multiple improvements & fixes. _Important:_
  `trezorWallet.register()` parameters were documented in the wrong order. The
  correct order is `(appUrl, email)`.

## 0.3.0 - 2019-10-19

### Added

- API: Add support for `changeTrust`.
- API: Add partial support for binary text memo. Text memo buffers containing
  NULL characters are still unsupported. See [trezor-firmware
  issue#610](https://github.com/trezor/trezor-firmware/issues/610)

### Changed

- API: Upgrade [trezor-connect] to 8.0.7. It fixes `bumpSequence` operation
  support.

### Fixed

- Logic: Fix handling of minTime|maxTime = 0.

## 0.2.0 - 2019-10-05

### Breaking

- API: `.disconnect()` is now synchronous.

### Added

- API: Add support for `trezor.connect(bipPath)`. `.connect()` now accepts
  either an account number as per Trezor definition, either a BIP path. For
  example, account 1 is `m/44'/148'/0'`.
- Logic: Add `setOptions/signer` support. (Thanks [@matejcik])

### Changed

- Documentation: Update [CHANGELOG.md]. Outline the fact that this is a beta
  release & that breaking changes may happen on minor updates (0.x.0).
- Documentation: Update [README.md].
- Logic: Skip validating pubkey on device. (Thanks [@matejcik])

### Fixed

- Logic: Fix account number handling. A mistake caused the wrong account to be
  used in some cases.

## 0.1.0 - 2019-09-28

Initial release.

[@matejcik]: https://github.com/matejcik
[readme.md]: https://cosmic.plus/#view:js-trezor-wallet
[changelog.md]: https://cosmic.plus/#view:js-trezor-wallet/CHANGELOG
[trezor-connect]: https://github.com/trezor/connect/blob/develop/CHANGELOG.md


src/trezor-tx.js
"use strict"
/**
 * Trezor Tx - A class that creates a Trezor-formatted transaction from any
 * format of transaction.
 */
const Buffer = require("@cosmic-plus/base/es5/buffer")
const StellarSdk = require("@cosmic-plus/base/es5/stellar-sdk")

const TxTransformer = require("./tx-transformer")

/* Definition */

class TrezorTransaction extends TxTransformer {
  constructor (transaction) {
    super(rules, transaction)
    delete this.network

    // Fees
    if (!this.fee) this.fee = 100 * this.operations.length
    else this.fee = Number(this.fee) / 10000000

    // Timebounds
    if (this.minTime != null || this.maxTime !== null) {
      this.timebounds = {
        minTime: this.minTime == null ? 0 : this.minTime,
        maxTime: this.maxTime == null ? 0 : this.maxTime
      }
      delete this.minTime
      delete this.maxTime
    }

    // Operations
    this.operations.forEach((op) => {
      if (op.type === "allowTrust") {
        op.assetType = op.assetCode.length <= 4 ? 1 : 2
        op.authorize = op.authorize ? 1 : 0
      } else if (op.type === "changeTrust") {
        if (!op.limit) op.limit = "9223372036854775807"
        op.line = op.asset
        delete op.asset
      } else if (op.type === "createPassiveSellOffer") {
        op.type = "createPassiveOffer"
      } else if (op.type === "manageSellOffer") {
        op.type = "manageOffer"
      } else if (op.type === "manageBuyOffer") {
        notSupported("operation manageBuyOffer")
      } else if (op.type === "pathPaymentStrictReceive") {
        op.type = "pathPayment"
      } else if (op.type === "pathPaymentStrictSend") {
        notSupported("operation pathPaymentStrictSend")
      } else if (op.type === "setOptions") {
        if (op.signer && op.signer.type === 0) {
          const keypair = StellarSdk.Keypair.fromPublicKey(op.signer.key)
          op.signer.key = keypair.rawPublicKey().toString("hex")
        }
      }
    })

    // DEBUG
    // eslint-disable-next-line no-console
    console.log("Trezor TX", JSON.stringify(this, null, 2))
  }
}

/* Rules */

const rules = {}

rules.amount = function (amount) {
  return Number(amount * 10000000).toFixed(0)
}

rules.asset = function (asset) {
  if (asset.code === "XLM" && !asset.issuer) return null
  asset.type = asset.code.length <= 4 ? 1 : 2
  return asset
}

rules.assetsArray = function (array) {
  return array.map(rules.asset)
}

rules.buffer = function (buffer) {
  if (!buffer.value) return ""

  if (buffer.type === "base64") {
    return Buffer.from(buffer.value, "base64").toString("hex")
  } else {
    return Buffer.from(buffer.value).toString("hex")
  }
}

rules.date = function (date) {
  return Number(new Date(date)) / 1000
}

rules.memo = function (memo) {
  if (memo.type === "text") {
    memo.type = 1
    memo.text = memo.value
  } else if (memo.type === "base64") {
    memo.type = 1
    memo.text = Buffer.from(memo.value, "base64").toString("utf8")
    if (memo.text.match("\u0000")) {
      notSupported("binary text memo including NULL characters")
    }
  } else if (memo.type === "id") {
    memo.type = 2
    memo.id = memo.value
  } else if (memo.type === "hash") {
    memo.type = 3
    memo.hash = memo.value
  } else if (memo.type === "return") {
    memo.type = 4
    memo.hash = memo.value
  }
  delete memo.value
  return memo
}

rules.price = function (price) {
  if (typeof price !== "object") return { n: Number(price), d: 1 }
  else return price
}

rules.signer = function (signer) {
  signer.key = signer.value
  delete signer.value
  if (signer.type === "key") signer.type = 0
  else if (signer.type === "tx") signer.type = 1
  else if (signer.type === "hash") signer.type = 2
  return signer
}

/* Helpers */
function notSupported (feature) {
  throw new Error(`This feature is not supported yet: ${feature}.`)
}

/* Export */
module.exports = TrezorTransaction


src/tx-transformer.js
"use strict"
/**
 * Tx Transformer
 *
 * Derives new formats from CosmicLink Transitional Description.
 */
const { CosmicLink, specs } = require("cosmic-lib")

/* Definition */

class TxTransformer {
  constructor (rules, tx) {
    const tdesc = new CosmicLink(tx).tdesc

    specs.transactionOptionalFields.forEach((field) => {
      const value = tdesc[field]
      if (value) this[field] = transform.field(rules, field, value)
    })
    this.operations = tdesc.operations.map((o) => transform.odesc(rules, o))
  }
}

/* Routines */

const transform = {}

transform.field = function (rules, field, value) {
  const type = specs.fieldType[field]
  return rules[type] ? rules[type](value) : value
}

transform.odesc = function (rules, odesc) {
  const result = {}
  for (let field in odesc) {
    result[field] = transform.field(rules, field, odesc[field])
  }
  return result
}

/* Export */
module.exports = TxTransformer


src/index.js
"use strict"
/**
 * This library is a convenient wrapper around the official
 * [TrezorConnect](https://github.com/trezor/connect) library.
 *
 * It provides a way to support Trezor devices with a few one-liners:
 *
 * ```js
 * // Step 1: Connect
 * await trezorWallet.connect()
 *
 * // Step 2: Get public key
 * const pubkey = trezorWallet.publicKey
 *
 * // Step 3: Sign
 * await trezorWallet.sign(transaction)
 *
 * // Extra: Event handlers
 * trezorWallet.onConnect = connectionHandler
 * trezorWallet.onDisconnect = disconnectionHandler
 * ```
 *
 * This library is browser-only.
 *
 * **Known limitations**
 *
 * A few operations & parameters are not supported yet. An explicit error
 * message will get thrown if you try to sign a transaction including them:
 *
 * - Binary text memo containing NULL characters,
 * - The `manageData` operation,
 * - The `manageBuyOffer` operation,
 * - The `pathPaymentStrictSend` operation.
 *
 * If you encounter an unexpected error, please [fill an
 * issue](https://github.com/cosmic-plus/js-trezor-wallet/issues/new/choose).
 *
 * @exports trezorWallet
 */
const trezor = exports

const TrezorConnect = require("trezor-connect").default

const Buffer = require("@cosmic-plus/base/es5/buffer")

const TrezorTx = require("./trezor-tx")

/* Configuration */
const BIP_PATH = "m/44'/148'"

/* Properties */

/**
 * PublicKey of the connected account.
 *
 * @member {String}
 */
trezor.publicKey = null

/**
 * Derivation path of the connected account. (default: `m/44'/148'/0'`)
 *
 * @member {String}
 */
trezor.path = null

/* Methods */

let connection

/**
 * Registers yourself in the Trezor Connect Manifest. This provides them the
 * ability to reach you in case of any required maintenance.
 *
 * This subscription is mandatory and the library register the Cosmic.plus
 * contact by default.
 *
 * @see [Trezor Connect Manifest](https://github.com/trezor/connect/blob/develop/docs/index.md#trezor-connect-manifest)
 *
 * @param {String} appUrl - Application URL.
 * @param {String} email - Developer email.
 */
trezor.register = function (appUrl, email) {
  TrezorConnect.manifest({ email, appUrl })
}
// Register this library by default.
trezor.register("@cosmic-plus/trezor-wallet", "mister.ticot@cosmic.plus")

/**
 * Waits for a connection with a Trezor wallet. If **account** is not provided,
 * account 1 is used. The library will stop listening for a connection if
 * `trezorWallet.disconnect()` is called.
 *
 * Once the connection is established, you can use `await
 * trezorWallet.connect(account)` again at any time to ensure the device is
 * still connected.
 *
 * When switching to another **account**, you can `await
 * trezorWallet.connect(new_account)` without prior disconnection.
 *
 * _Note:_ To stay consistent with the way Trezor numbers accounts, **account**
 * starts at 1 (derivation path: `m/44'/148'/0'`).
 *
 * @async
 * @param {Number|String} [account=1] - Either an account number (starts at 1)
 * or a derivation path (e.g: `m/44'/148'/0'`).
 */
trezor.connect = async function (account = 1) {
  let path = account
  if (typeof account === "number") {
    if (account < 1) throw new Error("Account number starts at 1.")
    path = `${BIP_PATH}/${account - 1}'`
  }

  // Connect & update data only when needed.
  if (trezor.path && trezor.path !== path) trezor.disconnect()
  if (!connection) connection = connect(path)
  return connection
}

async function connect (path) {
  // eslint-disable-next-line no-console
  console.log("Attempting Trezor connection...")
  connection = TrezorConnect.stellarGetAddress({
    path,
    showOnTrezor: false
  })
  const result = await connection

  if (result.success) {
    trezor.path = path
    trezor.publicKey = result.payload.address
    onConnect()
  } else {
    throw new Error(result.payload.error)
  }
}

/**
 * Prompts the user to accept **transaction** using the connected account of
 * their Trezor device.
 *
 * If the user accepts, adds the signature to **transaction**. Else, throws an
 * error.
 *
 * @async
 * @param {Transaction} transaction - A StellarSdk Transaction
 */
trezor.sign = async function (transaction) {
  if (!trezor.publicKey) throw new Error("No Trezor connected.")

  const result = await TrezorConnect.stellarSignTransaction({
    path: trezor.path,
    networkPassphrase: transaction.networkPassphrase,
    transaction: new TrezorTx(transaction)
  })
  if (!result.success) throw new Error(result.payload.error)

  const signature = Buffer.from(result.payload.signature, "hex")
  transaction.addSignature(trezor.publicKey, signature.toString("base64"))

  // DEBUG
  // eslint-disable-next-line no-console
  console.log("signed", transaction)
}

/**
 * Close the connection with the Trezor device, or stop waiting for one in case
 * a connection has not been established yet.
 */
trezor.disconnect = function () {
  try {
    // Try to close iframe. TODO: find a more reliable method
    TrezorConnect.cancel()
  } catch (error) {
    null
  }
  onDisconnect()
}

function reset () {
  connection = null
  const fields = ["path", "publicKey"]
  fields.forEach((name) => trezor[name] = null)
}

/**
 * Connects the first unused account.
 *
 * _Note:_ merged accounts are considered as used.
 *
 * @param {String|Server} [horizon=https://horizon.stellar.org] - The
 * Horizon server where to check for account existence. It can be either an URL
 * or a _StellarSdk.Server_ object.
 */
trezor.newAccount = async function (horizon = "https://horizon.stellar.org") {
  const accounts = await trezor.scan({
    horizon,
    attempts: 1,
    includeMerged: true
  })
  const last = accounts[accounts.length - 1]
  await trezor.connect(last ? last.account + 1 : 1)
}

/**
 * Scans the Trezor device for accounts that exist on **params.horizon**. The
 * scanning stops after encountering **params.attempts** unused accounts.
 *
 * Merged accounts are considered as existing accounts and will reset the
 * **params.attempts** counter when encountered.
 *
 * Returns an _Array_ of _Objects_ containing `account` number, `publicKey`,
 * `path`, and `state` (either `"open"` or `"merged"`).
 *
 * @param {Object} [params] - Optional parameters.
 * @param {String|Server} [params.horizon=https://horizon.stellar.org] - The
 * Horizon server where to check for account existence. It can be either an URL
 * or a _StellarSdk.Server_ object.
 * @param {Number} [params.attempts=3] - The number of empty accounts before
 * scanning stops.
 * @param {Boolean} [params.includeMerged=false] - List merged accounts as well.
 * @returns {Array}
 */
trezor.scan = async function ({
  horizon = "https://horizon.stellar.org",
  attempts = 3,
  includeMerged = false
}) {
  if (typeof horizon === "string") {
    const { Server } = require("@cosmic-plus/base/es5/stellar-sdk")
    horizon = new Server(horizon)
  }

  const accounts = []
  let miss = 0,
    index = 1,
    payload = []

  while (miss < attempts) {
    if (payload.length < index) {
      payload = payload.concat(await trezor.getPublicKeys(index, attempts * 5))
    }
    const current = payload[index - 1]

    // effectForAccount let us know about merged account as well.
    const callBuilder = horizon.effects().forAccount(current.publicKey)
    const response = await callBuilder
      .order("desc")
      .limit(1)
      .call()
      .catch(() => {})

    if (response) {
      miss = 0
      const latest = response.records[0]
      current.state = latest.type === "account_removed" ? "merged" : "open"

      if (includeMerged || current.state === "open") {
        accounts.push(current)
      }
    } else {
      miss++
    }

    index++
  }

  return accounts
}

/**
 * Request multiple public keys from the Trezor device. The request will return
 * **length** accounts, starting by **start** (minimum 1).
 *
 * Returns an _Array_ of _Objects_ with properties `account`, `publicKey`, and
 * `path`.
 *
 * @param {Number} [start=1] - Starting account number
 * @param {Number} [length=1] - Number of account to be listed.
 * @return {Array}
 */
trezor.getPublicKeys = async function (start = 1, length = 1) {
  const bundle = []

  for (let account = start; account < start + length; account++) {
    bundle.push({
      path: trezor.connect.path(account),
      showOnTrezor: false
    })
  }

  const response = await TrezorConnect.stellarGetAddress({ bundle })
  return response.payload.map((entry, index) => ({
    account: start + index,
    publicKey: entry.address,
    path: entry.serializedPath
  }))
}

/* Events */

TrezorConnect.on("DEVICE_EVENT", (event) => {
  // eslint-disable-next-line no-console
  console.log("DEVICE_EVENT", console.log(event))
  if (event.type === "device-disconnect") onDisconnect()
  else if (event.type === "device-connect") onConnect()
})

/**
 * _Function_ to execute on each connection.
 *
 * @category event
 * @member {Function}
 */
trezor.onConnect = null
function onConnect () {
  // eslint-disable-next-line no-console
  console.log("Trezor connected")
  if (typeof trezor.onConnect === "function") trezor.onConnect()
}

/**
 * _Function_ to execute on each disconnection.
 *
 * @category event
 * @member {Function}
 */
trezor.onDisconnect = null
function onDisconnect () {
  // eslint-disable-next-line no-console
  console.log("Trezor disconnected")
  reset()
  if (typeof trezor.onDisconnect === "function") trezor.onDisconnect()
}
